lib/internal/test_runner/test.js

```diff
@@ -0,0 +1,381 @@
+      // - Only called if there are results to report in the correct order.
+      // - Guaranteed to only be called a maximum of once per call to
+      //   processReadySubtestRange().
+      canSend = canSend || this.isClearToSend();
+
+      if (!canSend) {
+        return;
+      }
+
+      // Report the subtest's results and remove it from the ready map.
+      subtest.finalize();
+      this.readySubtests.delete(i);
+    }
+  }
+
+  createSubtest(name, options, fn) {
+    if (typeof name === 'function') {
+      fn = name;
+    } else if (name !== null && typeof name === 'object') {
+      fn = options;
+      options = name;
+    } else if (typeof options === 'function') {
+      fn = options;
+    }
+
+    if (options === null || typeof options !== 'object') {
+      options = ObjectCreate(null);
+    }
+
+    let parent = this;
+
+    // If this test has already ended, attach this test to the root test so
+    // that the error can be properly reported.
+    if (this.endTime !== null) {
+      while (parent.parent !== null) {
+        parent = parent.parent;
+      }
+    }
+
+    const test = new Test({ fn, name, parent, ...options });
+
+    if (parent.waitingOn === 0) {
+      parent.waitingOn = test.testNumber;
+    }
+
+    if (this.endTime !== null) {
+      test.fail(
+        new ERR_TEST_FAILURE(
+          'test could not be started because its parent finished',
+          kParentAlreadyFinished
+        )
+      );
+    }
+
+    ArrayPrototypePush(parent.subtests, test);
+    return test;
+  }
+
+  cancel() {
+    if (this.endTime !== null) {
+      return;
+    }
+
+    this.fail(
+      new ERR_TEST_FAILURE(
+        'test did not finish before its parent and was cancelled',
+        kCancelledByParent
+      )
+    );
+    this.cancelled = true;
+  }
+
+  fail(err) {
+    if (this.error !== null) {
+      return;
+    }
+
+    this.endTime = hrtime();
+    this.passed = false;
+    this.error = err;
+  }
+
+  pass() {
+    if (this.endTime !== null) {
+      return;
+    }
+
+    this.endTime = hrtime();
+    this.passed = true;
+  }
+
+  skip(message) {
+    this.skipped = true;
+    this.message = message;
+  }
+
+  todo(message) {
+    this.isTodo = true;
+    this.message = message;
+  }
+
+  diagnostic(message) {
+    ArrayPrototypePush(this.diagnostics, message);
+  }
+
+  start() {
+    // If there is enough available concurrency to run the test now, then do
+    // it. Otherwise, return a Promise to the caller and mark the test as
+    // pending for later execution.
+    if (!this.parent.hasConcurrency()) {
+      const deferred = createDeferredPromise();
+
+      deferred.test = this;
+      this.parent.addPendingSubtest(deferred);
+      return deferred.promise;
+    }
+
+    return this.run();
+  }
+
+  async run() {
+    this.parent.activeSubtests++;
+    this.startTime = hrtime();
+
+    try {
+      await this.runInAsyncScope(this.fn, null, new TestContext(this));
```

Comment:
```suggestion
      const ctx = new TestContext(this);
      await this.runInAsyncScope(this.fn, ctx, ctx);
```
To make this work:
```js
test(function myTest() {
  this.skip();
});
```

Reply:
I made this change, but I don't really like it. I don't think we should encourage using `this`. Mocha actually discourages using arrow functions in their docs.